### DISCUSSION TITLE
FTP port forward not working

### DISCUSSION THREAD

**Discussion by markaltmaier** [community/community#136746] [Codespaces]
How can I port forward the ftp tcp port on codespace terminal. My nodejs app uses ftp instances to connect to apps like filezilla, and I want to test the connection. However, it doesn't work when I port forward the port where it will connect on 3003 and copy the link, or the server ip both will not work?

How do I solve this issue?

**Comment by ferid333** [comment]
Hello !

Port forwarding FTP on GitHub Codespaces can be tricky because FTP requires more than just a single port to function properly. FTP typically uses two ports: 

- **Port 21** for the command (control) channel.
- **A range of ports** for the data channel.

Here’s how you can troubleshoot and solve this issue:

1. **Switch to Passive Mode:**  
   FTP has two modes: active and passive. Passive mode is generally easier to work with when dealing with port forwarding because the client initiates both connections. Make sure your FTP client (like FileZilla) and your Node.js app are configured to use passive mode.

2. **Port Forward the Data Port Range:**  
   Besides forwarding the command port (usually 21 or 3003 in your case), you also need to forward the range of ports used for the data connection. FTP in passive mode will use a range of ports, so you need to ensure those are forwarded in your Codespace settings. This might involve modifying your Node.js app to restrict the passive port range to something manageable, like 3004-3010, and then port forwarding those ports as well.

3. **Use FTP Over SFTP:**  
   If possible, consider using SFTP instead of FTP. SFTP only requires a single port (usually 22) and is much easier to handle in environments like Codespaces.

4. **Check Firewall Settings:**  
   Ensure that the Codespace or any network firewalls aren’t blocking the required ports. Since Codespaces are hosted environments, this should generally be fine, but it’s something to be aware of.

5. **Test with Local FTP Server:**  
   To narrow down the issue, try running an FTP server locally on your machine and test the connection with your Node.js app. This will help determine if the issue is with Codespaces or with your setup in general.

6. **Review Logs:**  
   Check the FTP server and client logs to get more insight into where the connection is failing. This might give you specific error messages that can point you in the right direction.

If none of these solutions work, consider reaching out to GitHub support or looking for issues specific to port forwarding and FTP in the Codespaces documentation.
